app/models/category.rb

```diff
@@ -1172,6 +1173,44 @@ def clear_site_cache
   def on_custom_fields_change
     clear_site_cache
   end
+
+  def self.reorder_categories!
+    categories =
+      if SiteSetting.fixed_category_positions
+        Category.order(:position)
+      else
+        Category
+          .left_outer_joins(:featured_topics)
+          .group("categories.id")
+          .order("max(topics.bumped_at) DESC NULLS LAST")
+      end.order(:id).pluck(:id, :parent_category_id)
+
+    # Start with an empty "root" category
+    ordered_categories = [[nil, nil]]
+
+    # Insert all subcategories after the parent category
+    ordered_categories.each_with_index do |(id, _), index|
+      if subcategory_ids = categories.filter { |_, parent_id| parent_id == id }.presence
+        ordered_categories.insert(index + 1, *subcategory_ids)
+      end
+    end
+
+    # Drop the "root" category
+    ordered_categories.shift
+
+    # Update all categories
+    subcategories = Hash.new(0)
+    depths = Hash.new(-1)
+    Category.transaction do
+      ordered_categories.each_with_index do |(id, parent_id), index|
+        Category.where(id: id).update_all(
+          position: index + 1,
+          position_to_parent: (subcategories[parent_id] += 1),
+          depth: (depths[id] = depths[parent_id] + 1),
+        )
+      end
+    end
```

Comment:
Executing a statement per category kills performance here.

```suggestion
    subcategories = Hash.new(0)
    depths = Hash.new(0)

    values =
      ordered_categories.each_with_index.map do |(id, parent_id), index|
        position_to_parent = (subcategories[parent_id] += 1) + 1
        depth = depths[id] = depths[parent_id] + 1
        "(#{id}, #{index + 1}, #{position_to_parent}, #{depth})"
      end.join(", ")

    DB.exec(<<~SQL)
      UPDATE categories
      SET position=v.position, position_to_parent=v.position_to_parent, depth=v.depth, updated_at=NOW()
      FROM (VALUES #{values}) as v(id, position, position_to_parent, depth)
      WHERE categories.id=v.id;
    SQL
```

With these two changes and 1,000,000 categories locally (3 levels with 100 subcategories per category), this executes in 30s for me locally.

Reply:
I knew that I could do some bulk UPDATE, but did not remember the syntax and this was easier. Will definitely have to improve this.